text/3649-macros-named-capture-groups.md

```diff
@@ -0,0 +1,797 @@
+$outer_a( $middle )*
+
+// Skipping a group level
+// Error: `$inner` must be contained within a `$middle` group, but it is within
+// `$outer_a`
+$outer_a( $inner( $x )* )*
+
+// No grouping at all
+// Error: `$x` must be within an `$inner` group, but it is not within any group
+$x
+```
+
+_TODO: if `foo[x0] foo[] foo[x2]` is captured and the expansion is
+`$foo( $x ),+`, should `x0, x2` be emitted of `x0, , x2` (extra comma)?
+Probably the first one._
+
+A possible relaxation of this rule is to allow groups or variables to expand to
+all captures within that level when not nested within the immediate parent.
+
+TBD whether this should be part of this RFC or a future possibility
+
+```rust
+// Expands to `x0 x1`
+$x
+
+// Expands to `x0 x1`
+$outer_a( $middle( $x )* )*
+```
+
+#### Expansion within non-parent groups
+
+If a group or metavariable is nested within a group that is not a capture
+parent, it should be repeated as many times as that group. It must still have
+its capture parents as expansion parents so as not to break other rules, but
+they need not be the immediate parents.
+
+In order to avoid edge cases with metavariable expressions, a group is not
+allowed to be nested within itself.
+
+```rust
+// Correct: prints `y0 y0 y0`
+// This is because the _entire_ expansion of `$outer_b` (one instance of $y)
+// is repeated once for each `$outer_a` (three instances)
+$outer_a( $outer_b( $y ) )
+
+// Correct: prints `ob[ oa[y0 i[x0 y0] oa[x1 y0]] oa[y0] oa[y0]]`.
+// Explanation:
+// - `root > outer_a > middle > inner > x` and `outer_b > y` ordering are both
+//   still respected, even though they are interleaved
+// - `outer_b` repeats once within root
+// - `outer_a` repeats three times within root, so repeats 3x within `outer_b`
+// - `inner` repeats `[2, 0, 0]` times within the `outer_a` instances. This
+//   drives how often `x` and `y` get repeated within that group.
+$outer_b( ob[$outer_a( oa[$y $middle( $inner( i[$x $y] ) )] )] )
+
+// Forbidden: `x` is missing parent `outer_a`
+$outer_b( $middle( $inner( $x $y ) ) ) )
+
+// Forbidden: group nesting within itself
+$outer_b( $outer_b( $y ) )
+$outer_b( $outer_b )
+```
+
+#### Single matches
+
+Capture groups must currently specify a kleene operator (`*`, `+`, or `?`) that
+determines if the group should match zero or more times, one or more times, or
+up to one time. This RFC will allow omitting the kleene operator to indicate
+that a group must be captured exactly once. That is, `$group(foo)` is a valid
+pattern (with current rules, `$(foo)` is forbidden).
+
+With this "exactly once" match there is no purpose in having a repetition token
+(e.g. the comma in `$(...),*`), so it must be omitted.
+
+#### Entire group expansion
+
+Since groups are named, it is now possible to write a group name to reemit its
+captured contents with no further expansion. This syntax uses the group name
+but omits the `(/* expansion pattern */)/* kleene */`:
```

Comment:
One potential source of ambiguity:

How would you express "I want to expand the entire named group, followed by some tokens in parentheses"?

I have a suggestion for how to solve that: `${$group}(tokens)`.